Hw 02 solution

kopervas opened the pull request:
The DTO layer has been implemented. The BookController, BookDto, BookMapper, MapperConfig, CreateBookRequestDto classes have been added.

kopervas committed:
Revert "Revert "1.0.0 First commit.""

This reverts commit 4871f167f807e6bbc0caa61e8b00c0b945f2ec05.

kopervas committed:
Revert "Revert "1.0.0 First commit.""

This reverts commit 4871f167f807e6bbc0caa61e8b00c0b945f2ec05.

kopervas committed:
1.0.1 Secont commit - check ci.yml.

kopervas committed:
1.0.1 Secont commit - check ci.yml.

kopervas committed:
1.0.2 Third commit - created the first model Book, BookRepository and BookService.

kopervas committed:
1.0.2 Third commit - created the first model Book, BookRepository and BookService.

kopervas committed:
1.0.3 Fourth commit. Bug fixes according to comments from Liliia Ponomarenko.

kopervas committed:
1.0.3 Fourth commit. Bug fixes according to comments from Liliia Ponomarenko.

kopervas committed:
1.0.4 Fifth commit. Bug fixes according to comments from Olena Bruiako.

kopervas committed:
1.0.4 Fifth commit. Bug fixes according to comments from Olena Bruiako.

kopervas committed:
1.0.5 Sixth commit. Bug fixes according to comments from Olena Bruiako.

kopervas committed:
1.0.5 Sixth commit. Bug fixes according to comments from Olena Bruiako.

kopervas committed:
1.0.6 Seventh commit. Added Controller, DTO, implemented MapStruct.

kopervas committed:
1.0.6 Seventh commit. Added Controller, DTO, implemented MapStruct.

kopervas committed:
1.0.7 Revert commit. Undo all changes made to the code in commit 1.0.6..

kopervas committed:
1.0.7 Revert commit. Undo all changes made to the code in commit 1.0.6..

kopervas committed:
Revert "1.0.7 Revert commit. Undo all changes made to the code in commit 1.0.6.."

This reverts commit 3eef526d0e0f9e713e93c4409651cec40258587c.

kopervas committed:
Revert "1.0.7 Revert commit. Undo all changes made to the code in commit 1.0.6.."

This reverts commit 3eef526d0e0f9e713e93c4409651cec40258587c.

kopervas committed:
Merge remote-tracking branch 'origin/hw-02-solution' into hw-02-solution

Elena-Bruyako commented on pom.xml:
```diff
@@ -47,6 +54,17 @@
+		<dependency>
+			<groupId>org.mapstruct</groupId>
+			<artifactId>mapstruct-processor</artifactId>
+			<version>1.5.5.Final</version>
```
mapstruct.version

Elena-Bruyako replied:
fix comment

Elena-Bruyako commented on pom.xml:
```diff
@@ -55,7 +73,32 @@
+						<path>
+							<groupId>org.mapstruct</groupId>
+							<artifactId>mapstruct-processor</artifactId>
+							<version>1.5.5.Final</version>
```
same here

liliia-ponomarenko replied:
the same

Elena-Bruyako commented on pom.xml:
```diff
@@ -55,7 +73,32 @@
+						<path>
+							<groupId>org.projectlombok</groupId>
+							<artifactId>lombok</artifactId>
+							<version>1.18.38</version>
```
lombok.version

Elena-Bruyako replied:
fix all previous comments

Elena-Bruyako commented on src/main/java/com/bookstore/controller/BookController.java:
```diff
@@ -0,0 +1,41 @@
+
+@RequiredArgsConstructor
+@RestController
+@RequestMapping(value = "/api/books")
```
add `server.servlet.context-path=/api` to application.properties
```suggestion
@RequestMapping("/books")
```

Elena-Bruyako commented on src/main/java/com/bookstore/controller/BookController.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    @GetMapping("/by-name")
+    public List<BookDto> getAllByName(@RequestParam String name) {
```
there is no such endpoint in task description

Elena-Bruyako commented on src/main/java/com/bookstore/dto/BookDto.java:
```diff
@@ -0,0 +1,25 @@
+@Setter
+@Data
+public class BookDto {
+    @Column(nullable = false)
```
this is dto, not an entity
remove constraints

Elena-Bruyako commented on src/main/java/com/bookstore/dto/BookDto.java:
```diff
@@ -0,0 +1,25 @@
+
+@Getter
+@Setter
+@Data
```
`@Data` already provide `@Getter` and `@Setter`

Elena-Bruyako commented on src/main/java/com/bookstore/repository/impl/BookRepositoryImpl.java:
```diff
@@ -3,46 +3,59 @@
-            transaction = session.beginTransaction();
-            session.save(book);
+        EntityTransaction transaction = null;
+        try (EntityManager entityManager = entityManagerFactory.createEntityManager()) {
```
don't use try with resources and transaction, session will close before catch block

Elena-Bruyako commented on src/main/java/com/bookstore/repository/impl/BookRepositoryImpl.java:
```diff
@@ -3,46 +3,59 @@
-            throw new DataProcessingException("Can`t get all books from DB", e);
+        try (EntityManager entityManager = entityManagerFactory.createEntityManager()) {
+            return entityManager.createQuery("SELECT u FROM Book u", Book.class).getResultList();
+        }
```
add catch block

Elena-Bruyako commented on src/main/java/com/bookstore/repository/impl/BookRepositoryImpl.java:
```diff
@@ -3,46 +3,59 @@
+        try (EntityManager entityManager = entityManagerFactory.createEntityManager()) {
+            Book book = entityManager.find(Book.class, id);
+            return Optional.ofNullable(book);
+        }
```
same here

Elena-Bruyako commented on src/main/java/com/bookstore/service/impl/BookServiceImpl.java:
```diff
@@ -1,24 +1,47 @@
-        return bookRepository.findAll();
+    public List<BookDto> findAll() {
+        return bookRepository.findAll().stream()
+                .map(bookMapper::toDto).toList();
```
```suggestion
        return bookRepository.findAll().stream()
                .map(bookMapper::toDto)
                .toList();
```

Elena-Bruyako commented on src/main/java/com/bookstore/service/impl/BookServiceImpl.java:
```diff
@@ -1,24 +1,47 @@
+    public BookDto findById(Long id) {
+        Book book = bookRepository.findById(id).orElseThrow(
+                () -> new EntityNotFoundException("Can`t find book by id " + id)
+                );
```
```suggestion
                () -> new EntityNotFoundException("Can`t find book by id " + id));
```

Elena-Bruyako reviewed (changes requested):
see comments

kopervas committed:
1.0.8 Correction of errors according to comments from Olena Bruiako.

liliia-ponomarenko commented on src/main/java/com/bookstore/repository/impl/BookRepositoryImpl.java:
```diff
@@ -1,48 +1,69 @@
+        try (EntityManager entityManager = entityManagerFactory.createEntityManager()) {
+            return entityManager.createQuery("SELECT u FROM Book u", Book.class).getResultList();
+        } catch (Exception e) {
+            throw new EntityNotFoundException("Can't find any books");
```
```suggestion
            throw new  DataProcessingException("Can't find any books");
```

liliia-ponomarenko commented on src/main/java/com/bookstore/repository/impl/BookRepositoryImpl.java:
```diff
@@ -1,48 +1,69 @@
+            return entityManager
+                    .createQuery("SELECT b FROM Book b WHERE lower(b.title) LIKE :name", Book.class)
+                    .setParameter("name", "%" + lowerCaseName + "%")
+                    .getResultList();
```
Do you use it?

liliia-ponomarenko commented on src/main/java/com/bookstore/service/impl/BookServiceImpl.java:
```diff
@@ -1,24 +1,47 @@
+                .stream().map(bookMapper::toDto)
+                .collect(Collectors.toList());
     }
 }
```
The same

liliia-ponomarenko reviewed (changes requested):
Be attentive and fix all previous comments ;)

kopervas committed:
1.0.9 Correction of errors according to comments from Liliia Ponomarenko.

Elena-Bruyako commented on pom.xml:
```diff
@@ -47,6 +54,17 @@
+		<dependency>
+			<groupId>org.mapstruct</groupId>
+			<artifactId>mapstruct</artifactId>
+			<version>1.5.5.Final</version>
```
mapstruct.version

liliia-ponomarenko replied:
```suggestion
			<version>${mapstruct.version}</version>
```

Elena-Bruyako commented on src/main/java/com/bookstore/controller/BookController.java:
```diff
@@ -0,0 +1,35 @@
+
+@RequiredArgsConstructor
+@RestController
+@RequestMapping(value = "/books")
```
```suggestion
@RequestMapping("/books")
```

liliia-ponomarenko replied:
not fixed

Elena-Bruyako commented on src/main/java/com/bookstore/repository/impl/BookRepositoryImpl.java:
```diff
@@ -1,48 +1,58 @@
+            return Optional.ofNullable(book);
         } catch (Exception e) {
-            throw new DataProcessingException("Can`t get all books from DB", e);
+            throw new EntityNotFoundException("Can`t find book by id " + id);
```
```suggestion
            throw new DataProcessingException("Can`t find book by id " + id);
```

Elena-Bruyako reviewed (changes requested):
see comments

kopervas committed:
1.1.0 Correction of errors according to comments from Olena Bruiako.

liliia-ponomarenko commented on pom.xml:
```diff
@@ -47,6 +54,17 @@
+		<dependency>
+			<groupId>org.mapstruct</groupId>
+			<artifactId>mapstruct-processor</artifactId>
+			<version>1.5.5.Final</version>
```
the same

liliia-ponomarenko commented on pom.xml:
```diff
@@ -55,7 +73,32 @@
+						<path>
+							<groupId>org.projectlombok</groupId>
+							<artifactId>lombok</artifactId>
+							<version>1.18.38</version>
```
not fixed

liliia-ponomarenko commented on pom.xml:
```diff
@@ -55,7 +73,32 @@
+						<path>
+							<groupId>org.projectlombok</groupId>
+							<artifactId>lombok-mapstruct-binding</artifactId>
+							<version>0.2.0</version>
```
```suggestion
							<version>&{lombok.mapstruct.binding.version}</version>
```

liliia-ponomarenko reviewed (changes requested):
Fix all previous comments ;)

liliia-ponomarenko commented on pom.xml:
```diff
@@ -73,9 +251,13 @@
 	</build>
-
 </project>
+-->
```
Remove commented code

liliia-ponomarenko reviewed (approved):
One comment, but fix it before merging ;)

kopervas committed:
1.2.0 Release of the task to the topic 'Spring Data JPA'. Used Lombok, Liquibase, implemented soft delete concept for the main entity.

kopervas committed:
1.2.0 Release of the task to the topic 'Spring Data JPA'. Used Lombok, Liquibase, implemented soft delete concept for the main entity.

Elena-Bruyako commented on src/main/java/com/bookstore/controller/BookController.java:
```diff
@@ -0,0 +1,43 @@
+        return bookService.findById(id);
+    }
+
+    @PostMapping
```
add status `HttpStatus.CREATED`

Elena-Bruyako commented on src/main/java/com/bookstore/dto/BookDto.java:
```diff
@@ -0,0 +1,14 @@
+import lombok.Data;
+
+@Data
+public class BookDto {
+    private String title;
```
```suggestion
public class BookDto {
    private Long id;
    private String title;
```

Elena-Bruyako commented on src/main/java/com/bookstore/model/Book.java:
```diff
@@ -7,9 +7,11 @@
 import lombok.Getter;
 import lombok.Setter;
 
+@Data
```
```suggestion
```

Elena-Bruyako commented on src/main/java/com/bookstore/model/Book.java:
```diff
@@ -30,4 +32,6 @@ public class Book {
     private String description;
     @Column
     private String coverImage;
+    @Column(name = "is_deleted", nullable = false)
```
remove `@Column` above `description` and `coverImage`
```suggestion
    @Column(nullable = false)
```

Elena-Bruyako commented on src/main/java/com/bookstore/repository/BookRepository.java:
```diff
@@ -2,9 +2,15 @@
-public interface BookRepository {
+@Repository
+public interface BookRepository extends JpaRepository<Book, Long> {
     Book save(Book book);
 
     List<Book> findAll();
+
+    Optional<Book> findById(Long id);
```
```suggestion
```

Elena-Bruyako commented on src/main/java/com/bookstore/repository/impl/BookRepositoryImpl.java:
```diff
@@ -1,48 +1,9 @@
-            throw new DataProcessingException("Can`t get all books from DB", e);
-        }
-    }
+public class BookRepositoryImpl {
```
remove `BookRepositoryImpl`

Elena-Bruyako commented on src/main/java/com/bookstore/service/impl/BookServiceImpl.java:
```diff
@@ -11,14 +16,40 @@
+    public BookDto update(Long id, UpdateBookRequestDto requestDto) {
+        Book book = bookRepository.findById(id)
+                .orElseThrow(() -> new RuntimeException("Book not found with id: " + id));
+
+        book.setTitle(requestDto.title());
+        book.setAuthor(requestDto.author());
+        book.setIsbn(requestDto.isbn());
+        book.setPrice(requestDto.price());
+        book.setDescription(requestDto.description());
+        book.setCoverImage(requestDto.coverImage());
+
```
add `void updateBookFromDto(CreateBookRequestDto book, @MappingTarget Book entity);` to `BookMapper`
```suggestion
        bookMapper.updateBookFromDto(requestDto, book);
```

Elena-Bruyako commented on src/main/resources/db/changelog/changes/add-is-deleted-to-books.yaml:
```diff
@@ -0,0 +1,14 @@
+databaseChangeLog:
+  - changeSet:
+      id: add-is-deleted-column
```
add `is_deleted` column to existing changelog

Elena-Bruyako commented on src/main/resources/db/changelog/db.changelog-master.yaml:
```diff
@@ -0,0 +1,47 @@
+                    nullable: false
+              - column:
+                  name: description
+                  type: VARCHAR(1024)
```
```suggestion
                  type: VARCHAR(255)
```

Elena-Bruyako commented on src/main/resources/db/changelog/db.changelog-master.yaml:
```diff
@@ -0,0 +1,47 @@
+                  type: VARCHAR(255)
+
+
+  - include:
+      file: db/changelog/changes/add-is-deleted-to-books.yaml
```
replace to `master` changelog file 
```suggestion
  - include:
      file: db/changelog/changes/add-is-deleted-to-books.yaml
```

Elena-Bruyako commented on src/main/resources/db/changelog/db.changelog-master.yaml:
```diff
@@ -0,0 +1,47 @@
+databaseChangeLog:
```
create separate changelog to create `books` table

Elena-Bruyako reviewed (changes requested):
Good job! Let's fix some comments

kopervas committed:
1.2.1 Release with bug fixes according to Olena Bruiako's comments.

kopervas committed:
1.2.2 Second release with bug fixes according to Olena Bruiako's comments.

kopervas committed:
1.2.3 Third release with bug fixes according to Olena Bruiako's comments.

kopervas committed:
1.2.4 Release with bug fixes according to Olena Bruiako's comments.
 Was:
   spring.jpa.hibernate.ddl-auto=none
 Became:
   spring.jpa.hibernate.ddl-auto=validate

kopervas committed:
1.3.0 An additional task to the chapter "Criteria Query in Spring Boot".

kopervas committed:
1.3.1 Errors correction according to comments from Olena Bruiako from 29.08.2025.

kopervas committed:
1.4.0 Implementation of GlobalExceptionHandler and Data Validation.
